README.md
VaultRE API Samples
==================

This repository contains sample cURL commands, PHP and Perl code for interacting with the VaultRE API.

Full API documentation (Swagger) is available at https://docs.api.vaultre.com.au/

The current version of the API is **1.3**.

Authentication Flow
==================

When you're ready to integrate with VaultRE, we will supply you with a unique *API Key*. Every request must include a request header `X-Api-Key` with your API Key. This is how we identify you as the partner accessing the API.

To access data for a VaultRE client, the client must generate and provide you with a unique *Access Token*. This should be provided with every request in the header `Authorization: Bearer <token>`.

Note that each *Access Token* has a set of scopes applied, as chosen by the VaultRE client. To view the scopes assigned to a token, you can call `GET /scopes` on the API.

Data Format
==================

The API transfers data in the JSON format.

All requests must be made over HTTPS.

Date/Time Format
==================

Generally, the API returns the following format for Date objects:

`YYYY-MM-DD`

e.g. `2016-04-28`

and the following format for DateTime objects:

`YYYY-MM-DDThh:mm:ss+z`

e.g. `2018-05-22T01:12:59+00:00`

Please note that all times are 24-hour and usually in the UTC timezone. You can submit DateTime objects in whatever timezone you prefer.

Rate Limiting
==================

All API Keys are rate limited to 10 requests per second and 10,000 requests per day. Therefore, you should cache data appropriately to avoid using up your allocated quota.

If you are retrieving property listings from the API to power a website, our recommendation is that you do not call the API "live" on each page load as this introduces the API as a direct dependency on your website performance and puts you at risk of exhausting your allocated quota. Rather, you should synchronise data periodically and cache it locally. 

Webhooks
==================

VaultRE offers outbound webhooks to interested integrators. In essence, VaultRE can send a small notification payload to your nominated endpoint when an event occurs in a client's VaultRE account. This can be a very efficient way of keeping two systems in sync without resorting to polling the API periodically for changes.

Versioning
==================

The JSON payload from various endpoints is subject to change at any time. Within a minor version (e.g. 1.0, 1.1, 1.2) fields can be added to a JSON payload but we will endeavour to never make breaking changes (e.g. removing a field or changing a field's data type). Between versions, breaking changes may be made. Your application should be able to handle new fields being added to the JSON payload arbitrarily.


Integrator Endpoints
==================

Special endpoints are available for integrators to access data at the integration level rather than the account level.

[More Information](INTEGRATORS.md)


Changelog
==================

A list of changes to the API including new endpoints, new fields, and changes in behaviour.

[More Information](CHANGELOG.md)


CHANGELOG.md
Changelog
==================
## v1.3


### 2022-11-17
- A new endpoint is available to retrieve single enquiries from the holding area (GET /enquiries/{id})
- POST requests on "/properties/{id}/activityLog" endpoint are now supported


### 2022-11-11
- The "internalMarketingLiveDate" field is now displayed on property life objects. This represents the date the listing was first marked for "Internal Marketing".


### 2022-11-03
- Property list endpoints can now be filtered by "letAgreed" attribute


### 2022-09-12
- Sale details for properties now include the net income to office field
- Office referrals are now also included in the sale details


### 2022-09-07
- The "furnished" object is now exposed on the lease property endpoints


### 2022-08-24
- The "isOurListing" field is now displayed on sold and sale property life objects


### 2022-06-24
- The "isNewHome" field is now displayed on property life objects


### 2022-06-03
- New endpoints are available for retrieving listing configurations and strata properties under a strata building record


### 2022-05-30
- Lease properties now include council rates (UK accounts only)
- Sale properties now include shared ownership details (UK accounts only)
- Leasehold details are now exposed on the single sale property endpoints (UK accounts only)
- A new endpoint is available to retrieve property rooms (UK accounts only)


### 2022-05-17
- Solicitors for a given property life can now be fetched.


### 2022-04-27
- Property files now include an "isPublic" attribute, which can also be filtered in the request.


### 2022-03-13
- Adverting schedules can now be fetched
- Adverting schedules assigned to a property can now be fetched
- A new endpoint to retrieve bulk "source of enquiry" mappings is now available
- Previously submitted property portal stats can now be retrieved


### 2022-03-08
- Adverting expense types can now be fetched
- Adverting transactions on a property ledger can now be fetched


### 2022-01-31
- A new endpoint is available to retrieve property offers (GET /propertyOffers)


### 2021-12-24
- All property endpoints can now filter by property type
- All property endpoints can now sort by land or floor area
- The /user endpoint now contains information about the teams to which the user is attached


### 2021-09-24
- The sold property endpoints can now optionally include fallen/crashed conditional (written) sales


### 2021-09-21
- Pre-distribution deductions are now included on sold endpoints
- The Gross Commission (less deductions) figure is now available on sold endpoints


### 2021-09-16
- The Council Tax Band field is now displayed on the property model (UK accounts only)


### 2021-08-18
- Sold properties can now be filtered by more fields (bed,bath,car,property type, sale price)


### 2021-08-11
- A new field "electronicTable" is now exposed on the property life files endpoint, indicating whether
  or not a file is exposed on the eTable for the given sale/lease.


### 2021-08-10
- The sold property endpoints can now optionally include conditional (written) sales


### 2021-07-29
- Deductions are now included on commission splits


### 2021-07-10
- Precincts can now be fetched (GET /precincts)
- Properties can now be filtered by precinct


### 2021-06-26
- Custom unsubscribe types per account are now supported
- A new endpoint, /types/customUnsubscribe, is now available to retrieve all unsubscribe types for the account
- The Contact model now returns the customUnsubscribe type array. This can also be set on POST/PUT operations.


### 2021-06-22
- A new endpoint, /contacts/dates,  is now available to retrieve all contact dates


### 2021-06-21
- Suburb objects now include an "nzDistrict" field


### 2021-05-31
- It is now possible to submit enquiries attached to a specific user rather than a property (e.g. from a "Contact Me" page)
- Enquiries can now also be filtered by a particular user (for user-attached enquiries)


### 2021-05-25
- releaseDate and isReleasedSale fields are now available in the SaleHistory object. These relate to release of funds in Victoria and New Zealand.


### 2021-05-04
- A new endpoint /openHomes is now available to retrieve open homes across multiple properties.


### 2021-04-22
- The "contactStaff" array is now included in the saleHistory object on a property.
- The "contactStaff" array is now included in the leaseHistory object on a property.


### 2021-04-16
- When uploading a file to the filing cabinet for a lease property, you can now specify to which tenancy record the file relates.


### 2021-04-14
- "publishedToWeb" date field is now available on sale properties. This represents the date a sale property was first published to the web.


### 2021-04-08
- "inserted" and "modified" fields are now available in the SaleHistory and LeaseHistory objects


### 2021-03-10
- A new endpoint, /properties/{id}/{salelease}/{lifeid}/portalIds is now available to retrieve known IDs of properties on portal sites


### 2021-03-02
- The /account/users endpoint now allows filtering by email address


### 2021-02-25
- A new endpoint is available to find all contact records merged in to a given contact (GET /contacts/{contactid}/merged)


### 2021-02-15
- The bulk contact notes endpoint can now be filtered by insertedSince and insertedBefore


### 2021-02-12
- /types/holdingArea Retrieve a list of holding area types
- /enquiries now accepts a "source" parameter. Pass in a holding area type API code to filter enquiries by type.


### 2021-02-08
- Sold properties can now be filtered by suburb
- Contact objects now display a smsUnsubscribeUrl field


### 2021-01-27
- It is now possible to add or update contact category groupings


### 2021-01-18
- Tasks can now be filtered by date inserted


### 2020-12-10
- The bulk current owner/purchaser/landlord/tenant endpoints now allow filtering by modify date
- The account postal address is now included in the Account model


### 2020-12-09

aggregator/client.py
import json
import requests
import jwt
import time
from uuid import uuid4
from datetime import datetime

SECRET = 'YOUR_SECRET_KEY'
CRMKEY = 'YOUR_CRM_KEY'

BASE_URL = 'https://aggregator.api.vaultre.com.au/api/v1.0'


def main():

    token = generate_jwt()

    headers = {
        'X-Api-Key': SECRET,
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ' + token,
    }

    call_staff(headers)
    call_appraisal(headers)
    call_listing(headers)
    call_unconditional(headers)
    call_settlement(headers)
    call_withdrawn(headers)


def call_api(path, headers, data):

    r = requests.post(BASE_URL + path, headers=headers, data=data)
    print(path + ' - ' + r.text)


def call_staff(headers):

    now = datetime.utcnow().replace(microsecond=0).isoformat()

    payload = {
        "transactionId": str(uuid4()),
        "transactionDate": now,
        "staffId": "900",
        "officeId": "272",
        "active": True,
        "firstName": "Matt",
        "lastName": "Healy",
        "position": "Salesperson",
        "mobile": "0400 000 000",
        "email": 'test@example.com',
    }

    data = json.dumps(payload)
    call_api('/staff', headers, data)


def call_appraisal(headers):

    now = datetime.utcnow().replace(microsecond=0).isoformat()

    payload = {
        "transactionId": str(uuid4()),
        "transactionDate": now,
        "appraisalId": "1",
        "propertyId": "1",
        "officeId": "272",
        "staff": [
            {"staffId": "900"},
        ],
        "appraisalDate": "2021-03-27",
        "appraisalPrice": 480000,
        "classification": "residential",
        "active": True,
        "address": {
            "unitNumber": "3",
            "streetNumber": "20",
            "street": "Fake Street",
            "suburb": "West Perth",
            "postcode": "6005",
            "state": "WA",
        },
    }

    data = json.dumps(payload)
    call_api('/appraisal', headers, data)


def call_listing(headers):

    now = datetime.utcnow().replace(microsecond=0).isoformat()

    payload = {
        "transactionId": str(uuid4()),
        "transactionDate": now,
        "appraisalId": "1",
        "propertyId": "1",
        "listingId": "1",
        "officeId": "272",
        "staff": [
            {"staffId": "900"},
        ],
        "auctionDate": "2021-03-27",
        "listingAuthorityDate": "2021-03-27",
        "auctionVenue": "On Site",
        "listingPrice": 450000,
        "classification": "residential",
        "active": True,
        "address": {
            "unitNumber": "3",
            "streetNumber": "20",
            "street": "Fake Street",
            "suburb": "West Perth",
            "postcode": "6005",
            "state": "WA",
        },
        "listingAuthorityType": "exclusive",
        "methodOfSale": "privateTreaty",
    }

    data = json.dumps(payload)
    call_api('/listing', headers, data)


def call_unconditional(headers):

    now = datetime.utcnow().replace(microsecond=0).isoformat()

    payload = {
        "transactionId": str(uuid4()),
        "transactionDate": now,
        "unconditionalId": "1",
        "propertyId": "1",
        "listingId": "1",
        "officeId": "272",
        "commissionSplits": [
            {
                "staffId": "900",
                "type": "lister",
                "amountExGST": "5000",
            },
            {
                "staffId": "901",
                "type": "seller",
                "amountExGST": "7000",
            },
        ],
        "unconditionalDate": "2021-03-27",
        "proposedSettlementDate": "2021-03-27",
        "salePrice": 465000,
        "grossCommissinExGST": 12000,
        "classification": "residential",
        "active": True,
        "address": {
            "unitNumber": "3",
            "streetNumber": "20",
            "street": "Fake Street",
            "suburb": "West Perth",
            "postcode": "6005",
            "state": "WA",
        },
        "soldType": "privateTreaty",
    }

    data = json.dumps(payload)
    call_api('/unconditional', headers, data)


def call_settlement(headers):

    now = datetime.utcnow().replace(microsecond=0).isoformat()

    payload = {
        "transactionId": str(uuid4()),
        "transactionDate": now,
        "unconditionalId": "1",
        "settlementId": "1",
        "propertyId": "1",
        "listingId": "1",
        "officeId": "272",
        "commissionSplits": [
            {
                "staffId": "900",
                "type": "lister",
                "amountExGST": "5000",
            },
            {
                "staffId": "901",
                "type": "seller",
                "amountExGST": "7000",
            },
        ],
        "unconditionalDate": "2021-03-27",
        "settlementDate": "2021-03-27",
        "salePrice": 465000,
        "grossCommissinExGST": 12000,
        "classification": "residential",
        "active": True,
        "address": {
            "unitNumber": "3",
            "streetNumber": "20",
            "street": "Fake Street",
            "suburb": "West Perth",
            "postcode": "6005",
            "state": "WA",
        },
        "soldType": "privateTreaty",
        "vendorPaidAdvertising": 1500,
        "agentPaidAdvertising": 100,
    }

    data = json.dumps(payload)
    call_api('/settlement', headers, data)


def call_withdrawn(headers):

    now = datetime.utcnow().replace(microsecond=0).isoformat()

    payload = {
        "transactionId": str(uuid4()),
        "transactionDate": now,
        "appraisalId": "1",
        "propertyId": "1",
        "listingId": "1",
        "withdrawnId": "1",
        "officeId": "272",
        "staff": [
            {"staffId": "900"},
        ],
        "withdrawnDate": "2021-03-27",
        "classification": "residential",
        "active": True,
        "address": {
            "unitNumber": "3",
            "streetNumber": "20",
            "street": "Fake Street",
            "suburb": "West Perth",
            "postcode": "6005",
            "state": "WA",
        },
        "vendorPaidAdvertising": 1500,
        "agentPaidAdvertising": 100,
    }

    data = json.dumps(payload)
    call_api('/withdrawn', headers, data)


def generate_jwt():

    token_payload = {
        'timestamp': int(time.time()),
        'crmKey': CRMKEY,
    }

    token = jwt.encode(token_payload, SECRET, algorithm="HS512")

    return token


if __name__ == "__main__":
    main()
